modules/transactions/src/main/java/org/apache/ignite/internal/tx/storage/state/rocksdb/TxStateRocksDbStorage.java

```diff
@@ -545,95 +492,74 @@ public CompletableFuture<Void> startRebalance() {
-                .thenAccept(unused -> {
-                    try (WriteBatch writeBatch = new WriteBatch()) {
-                        writeBatch.deleteRange(partitionStartPrefix(), partitionEndPrefix());
-                        writeBatch.delete(lastAppliedIndexAndTermKey);
+        try (WriteBatch writeBatch = new WriteBatch()) {
+            writeBatch.deleteRange(partitionStartPrefix(), partitionEndPrefix());
+            writeBatch.delete(lastAppliedIndexAndTermKey);
 
-                        db.write(writeOptions, writeBatch);
+            db.write(writeOptions, writeBatch);
 
-                        lastAppliedIndex = 0;
-                        lastAppliedTerm = 0;
-                        persistedIndex = 0;
+            lastAppliedIndex = 0;
+            lastAppliedTerm = 0;
+            persistedIndex = 0;
 
-                        state = StorageState.RUNNABLE;
-                    } catch (Exception e) {
-                        throw new IgniteInternalException(
-                                TX_STATE_STORAGE_REBALANCE_ERR,
-                                IgniteStringFormatter.format(
-                                        "Failed to clear storage for partition {} of table {}",
-                                        partitionId,
-                                        getTableName()
-                                ),
-                                e
-                        );
-                    }
-                });
+            state.set(StorageState.RUNNABLE);
+        } catch (Exception e) {
+            throw new IgniteInternalException(
+                    TX_STATE_STORAGE_REBALANCE_ERR,
+                    IgniteStringFormatter.format("Failed to abort rebalance: [{}]", createStorageInfo()),
+                    e
+            );
+        }
+
+        return completedFuture(null);
     }
 
     @Override
     public CompletableFuture<Void> finishRebalance(long lastAppliedIndex, long lastAppliedTerm) {
-        CompletableFuture<Void> rebalanceFuture = (CompletableFuture<Void>) REBALANCE_FUTURE.getAndSet(this, null);
-
-        if (rebalanceFuture == null) {
-            throw new IgniteInternalException(TX_STATE_STORAGE_REBALANCE_ERR, "Rebalancing has not started");
+        if (state.get() != StorageState.REBALANCE) {
+            throw new IgniteInternalException(
+                    TX_STATE_STORAGE_REBALANCE_ERR,
+                    IgniteStringFormatter.format("Rebalancing has not started: [{}]", createStorageInfo())
+            );
         }
 
-        return rebalanceFuture
-                .thenAccept(unused -> {
-                    try (WriteBatch writeBatch = new WriteBatch()) {
-                        writeBatch.put(lastAppliedIndexAndTermKey, indexAndTermToBytes(REBALANCE_IN_PROGRESS, REBALANCE_IN_PROGRESS));
+        try (WriteBatch writeBatch = new WriteBatch()) {
+            writeBatch.put(lastAppliedIndexAndTermKey, indexAndTermToBytes(REBALANCE_IN_PROGRESS, REBALANCE_IN_PROGRESS));
```

Comment:
Why do we write special values `<REBALANCE_IN_PROGRESS, REBALANCE_IN_PROGRESS>` instead of real index+term here?

Reply:
Fix it. thanks!